rfcs/2021-08-29-8381-vrl-iteration-support.md

```diff
@@ -0,0 +1,756 @@
+```
+
+## Rationale
+
+Iteration unlocks solutions to many remapping scenarios we currently don't
+support. Not implementing this RFC would hold VRL back, and prevent operators
+with more complex use-cases from using Vector with VRL to achieve their goals.
+
+By adding iteration, we unlock the capability to resolve almost all use-cases in
+the future by introducing more iteration-based functions.
+
+## Drawbacks
+
+- It adds more complexity to the language.
+- There are potential performance foot guns when iterating over large
+  collections.
+- The parser and compiler have to become more complex to support this use-case.
+
+## Prior Art
+
+TODO
+
+## Alternatives
+
+### For-Loop
+
+A different approach to iteration is to use a built-in syntax for-loop:
+
+```coffee
+for (key, _value) in . {
+  key = upcase(key)
+}
+```
+
+The biggest strength of this approach is the simplicity of the syntax, and the
+familiarity with many other languages that have for-loops.
+
+It's relevant to mention that this solution also still needs lexical-scoping
+implemented, to avoid "leaking" the values of the `key` and `value` variables
+outside of the loop.
+
+One problem with this approach is that recursive iteration (accessing nested
+object fields) isn't possible, unless we add another special syntax (e.g.
+`recursive for (.., ..) in . {}`). This adds more surface-level syntax and
+removes some of its familiarity, making it a less attractive solution.
+
+An additional problem is that the `key` and `value` variables become "special",
+in that, even though it _appears_ that they aren't used after assignment, the
+`for-loop` expression would actually update the object key after each iteration
+in the loop.
+
+While this is technically the same problem we had to solve in the function-based
+solution, applying that same solution to a `for-loop` again makes it look less
+like for-loops in other languages, defeating one of the strengths of this
+approach:
+
+```coffee
+for (key, value) in . {
+  key = upcase(key)
+
+  (key, value)
+}
+```
+
+A solution to the magic-variable problem would be to allow dynamic paths, and
+have operators directly assign to those paths:
+
+```coffee
+for (key, _value) in . {
+  .[upcase(key)] = value
+}
+```
+
+This solves one problem, but introduces another: using `.<path>` always starts
+at the root of the target. Given the following example:
+
+```json
+{ "foo": { "bar": true } }
+```
+
+How would we use dynamic paths in a recursive for-loop?
+
+```coffee
+recursive for (key, value) in . {
+  .[upcase(key)] = value
+}
+```
+
+Because key is `"foo"` and then `"bar"`, you would end up with:
+
+```json
+{ "FOO": true, "BAR": true }
+```
+
+Which is not the expected outcome.
+
+This could be solved by making `.` relative in the for-loop, but that's a major
+shift from the current way VRL works, requires a new way of accessing the root
+object if you can't use `.`, and goes against the rules as laid out in the
+[design document][doc].
```

Comment:
Reminds me of the pain in Helm templating where you use `$` instead for "local" things, which I don't love.

Reply:
:nod: yeah.